services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/DownloadDirectoryHelper.java

```diff
@@ -110,78 +112,167 @@ private void doDownloadDirectory(CompletableFuture<CompletedDirectoryDownload> r
+                         .subscribe(new S3ObjectSubscriber(downloadDirectoryRequest,
+                                                           returnFuture,
+                                                           downloadFileFunction,
+                                                           DEFAULT_DOWNLOAD_DIRECTORY_MAX_CONCURRENCY));
+    }
+
+    @SdkInternalApi
+    static class S3ObjectSubscriber implements Subscriber<S3Object> {
+        private static final Logger log = Logger.loggerFor(S3TransferManager.class);
+        private final List<CompletableFuture<CompletedFileDownload>> currentBuffer;
+        private final Collection<FailedFileDownload> failedFileDownloads;
+        private final DownloadDirectoryRequest request;
+        private Subscription subscription;
+        private final CompletableFuture<CompletedDirectoryDownload> returnFuture;
+        private final Function<DownloadFileRequest, FileDownload> downloadFileFunction;
+        private final int bufferSize;
+
+        S3ObjectSubscriber(DownloadDirectoryRequest request,
+                           CompletableFuture<CompletedDirectoryDownload> returnFuture,
+                           Function<DownloadFileRequest, FileDownload> downloadFileFunction,
+                           int concurrentDownload) {
+            this.failedFileDownloads = new ConcurrentLinkedQueue<>();
+            this.currentBuffer = new ArrayList<>();
+            this.request = request;
+            this.returnFuture = returnFuture;
+            this.downloadFileFunction = downloadFileFunction;
+            this.bufferSize = concurrentDownload;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            if (subscription == null) {
+                throw new NullPointerException("Subscription must not be null");
+            }
+
+            if (this.subscription != null) {
+                subscription.cancel();
+                return;
+            }
+
+            this.subscription = subscription;
+            subscription.request(1);
+        }
+
+        @Override
+        public void onNext(S3Object s3Object) {
+            log.debug(() -> "Processing s3Object key: " + s3Object.key());
+            CompletableFuture<CompletedFileDownload> singleFileFuture = downloadSingleFile(request,
+                                                                                           failedFileDownloads,
+                                                                                           s3Object);
+            currentBuffer.add(singleFileFuture);
+
+            if (currentBuffer.size() == bufferSize) {
+                CompletableFuture.allOf(currentBuffer.toArray(new CompletableFuture[0]))
+                                 .whenComplete((response, throwable) -> {
+                                     subscription.request(1);
+                                     currentBuffer.clear();
+                                 });
```

Comment:
I think this buffer implementation may result in creating bursts of demand, rather than a smooth continuous demand. E.g., once the buffer is full, we will entirely cease to list objects, until all N objects finish, at which point we will resume listing (but be briefly stuck awaiting a list response). This may be partially offset by the fact that List is a paginated operation, so we only really need 1 call to resume meaningful work, but there will still be a small period of delay that we could probably avoid...

So rather than buffering in batches, could we possibly use a `Queue` that we continuously write to and read from, and only cease requesting more objects when the queue reaches a max size?

Also, would it be possible to expose this functionality via a convenience method like [`SdkPublisher#buffer`](https://github.com/aws/aws-sdk-java-v2/blob/8c79c6b1ecc313e535065e86abaf0d43c12437c5/core/sdk-core/src/main/java/software/amazon/awssdk/core/async/SdkPublisher.java#L98-L107)?

Reply:
Fancy! The implementation would be more complicated though. I'm not sure if the benefit justifies the efforts because there will not be any delay for the case where there are still s3Objects from the previous ListObjectsV2Response in memory, and even if we need to call listObjectsV2, the latency seems to be trivial in the process of downloading multiple objects. 

> So rather than buffering in batches, could we possibly use a Queue that we continuously write to and read from, and only cease requesting more objects when the queue reaches a max size?

It seems `size()` is not a constant-time operation in the concurrent queue implementations and we probably don't want to use blocking queue or add locks before inserting and removing. There are ways to make it work, but more complex.

 > Also, would it be possible to expose this functionality via a convenience method like [SdkPublisher#buffer](https://github.com/aws/aws-sdk-java-v2/blob/8c79c6b1ecc313e535065e86abaf0d43c12437c5/core/sdk-core/src/main/java/software/amazon/awssdk/core/async/SdkPublisher.java#L98-L107)?

I'll see if I can generalize it.

Reply:
Good point that most of the time there will be a remaining/partial list response still in memory. Also agree that the list delay would be mostly amortized in the context of bulk object transfers, but I think it's still something we could consider optimizing for.

I think another, perhaps more concerning scenario, is when the 100th download takes a very long time to download, and all other 99 downloads have finished, there will only be 1 in-flight transfer until the 100th completes.

Reply:
Yeah, that is an interesting scenario. Good point, I'm convinced. I'll make the change.